Implement AttentionGate mechanism

bil-y opened the pull request:
- Implemented the `AttentionGate` mechanism from https://arxiv.org/abs/1804.03999
- Created `AttnGateUpBlock` which can be integrated into a UNet in place of `UpBlock`.
- Added tests

mstadelmann commented:
Not sure about the resampling question, I'll have to take a closer look to check what exactly is happening there.
I added it into a new issue so that we don't forget about it: https://github.com/CAIIVS/chuchichaestli/issues/11

mstadelmann reviewed (dismissed):
Looking good.
(Have not yet had the time to run experiments on it, but ok for me to merge. Will run a test with it next week.)

danielbarco commented:
The monai implementation seems to perform the attention gate operation on the higher layer (x_l/ skip connection dimensionality). As far as I can see, similar to your code. So I guess, both is possible...

bil-y commented:
> hi @bil-y, good job on the implementation. Had to reread the paper several times. concerning the resampling correct me if I am wrong here are my thoughts: subscript g (gating signal) refers to the coarser/lower layer, while x^l (skip connection) refers to finer/higher layer. The operations in the attention gate are performed on the coarser level (cf. figure 2). Then we have the operations in the attention gate (addition, ReLu, Psi, Sig) all on the dimensionality of the coarser level. Then the resampler upscales (trilinear interpolation) the dimensionality to a finer layer. There is then a tensor product with the skip connection. If you check figure 1 you will see that e.g. the bottle neck F_4 x H_4 x W_4 x D_4 has two path ways 1. with an upsampling and 2. through the attention gate (where it must be upsampled as well). Let me know if you agree. Figure 2 is definitely confusing as g is on the top, while it represents the gating signal from the lower layer...

Thanks! Now that I've switched `g` and `x_l`, the "channel math" actually makes sense. 

I'm still unsure about whether we actually need upscaling. The dimensions fit the way they are, and the ResNet blocks perform upscaling of their own, so I don't see the benefit. It appears that the Monai implementations need them because they don't have any other upscaling layers anywhere else. 🤷 Thoughts?

danielbarco commented:
@bil-y Thank you for the update. I was looking at your code especially https://github.com/CAIIVS/chuchichaestli/blob/8037fd6c345d6359b58925e78c57a5a49ffc6eb9/src/chuchichaestli/models/attention/attention_gate.py#L201 . I was playing around with the code a little, but would have to test more thoroughly. In the paper they mention, that using a coarser layer to guide the skip connection is useful (cf. figure 3a). Also I assume performing the attention gate operations on the coarser layer would be more memory efficient. I imagined, that the input to the attention_gate would be e.g. torch.Size([1, 512, 8, 8]) and torch.Size([1, 256, 16, 16]) for g and x_l respectively. Is this the way the current implementation works? 
I think what we want in the end, is an implementation equivalent to the one published in the original paper (with deviations, documented and described) and also to benchmark the performance of this attention UNet with that of Monai

bil-y commented:
> @bil-y Thank you for the update. I was looking at your code especially
> 
> https://github.com/CAIIVS/chuchichaestli/blob/8037fd6c345d6359b58925e78c57a5a49ffc6eb9/src/chuchichaestli/models/attention/attention_gate.py#L201
> . I was playing around with the code a little, but would have to test more thoroughly. In the paper they mention, that using a coarser layer to guide the skip connection is useful (cf. figure 3a). Also I assume performing the attention gate operations on the coarser layer dimensions would be more memory efficient. I imagined, that the input to the attention_gate would be e.g. torch.Size([1, 512, 8, 8]) and torch.Size([1, 256, 16, 16]) for g and x_l respectively. My thinking is, that taking g from the coarser level (prior to upsampling) should be beneficial (at least according to the paper). Is this the way the current implementation works? Not sure I managed to reproduce this with your code, but may be mistaken.
> I think what we want in the end, is an implementation equivalent to the one published in the original paper (or with deviations that are documented and described) and also to benchmark the performance of this attention UNet with that of Monai.

I've moved the implementation closer to what the authors published in their own GitHub: https://github.com/ozan-oktay/Attention-Gated-Networks/tree/master. I now consequently use the lower-level outputs pre-upscaling as gating vectors. I've also implemented their "grid upscaling".

danielbarco reviewed (approved):
Lets test the Monai AttentionUnet against chuchichästli.

bil-y commented on tests/test_attention_gate.py:
```diff
@@ -0,0 +1,69 @@
+
+    # Create attention gate module
+    attention_gate = AttentionGate(
+        dimension=dimension, num_channels_g=3, num_channels_x=6, num_channels_inter=3
```
Probably a left-over from when it mixed up g and x. I've changed it now, but it doesn't change the outcome of the test.

danielbarco reviewed (approved):
lets test
